src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ReliableMessagingHelpers.cs

```diff
@@ -2,11 +2,215 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
 using System.Runtime;
+using System.ServiceModel.Security;
+using System.Threading;
 using System.Threading.Tasks;
+using System.Xml;
 
 namespace System.ServiceModel.Channels
 {
+    internal delegate Task OperationWithTimeoutAsyncCallback(TimeSpan timeout);
+
+    internal sealed class Guard
+    {
+        private TaskCompletionSource<object> _tcs;
+        private int _currentCount = 0;
+        private int _maxCount;
+        private bool _closed;
+        private object _thisLock = new object();
+
+        public Guard() : this(1) { }
+
+        public Guard(int maxCount)
+        {
+            _maxCount = maxCount;
+        }
+
+        public void Abort()
+        {
+            _closed = true;
+        }
+
+        public async Task CloseAsync(TimeSpan timeout)
+        {
+            lock (_thisLock)
+            {
+                if (_closed)
+                    return;
+
+                _closed = true;
+
+                if (_currentCount > 0)
+                    _tcs = new TaskCompletionSource<object>();
+            }
+
+            if (_tcs != null)
+            {
+                try
+                {
+                    if (!await _tcs.Task.AwaitWithTimeout(timeout))
```

Comment:
Do you not use `ConfigureAwait` in general? We always use `ConfigureAwait(false)` to prevent sync context capture on all awaits, but if you're not using it consistently then it doesn't really matter.

Reply:
We take a different approach. Because all our public api methods are very large composite operations (doing lots of things including multiple IO operations), we know we will always go async at some point before returning to the caller. Littering our code everywhere with ConfigureAwait is risky because if we miss just one usage, we could have a very difficult to track down bug. Instead at public api entry I have a helper method/class that I await on. What it does is check if a non-default sync context or non-default Task scheduler is being used, and if it is, immediately yields the thread and continues on the default sync context. If the default sync context is being used, the await continues synchronously. This ensures we hop threads if there's a risk of deadlocking, but is a no-op if there's no risk. This only makes sense because the amount of work a single call does is massive compared with something like Stream.ReadAsync.

Reply:
Interesting, makes sense.

It seems like you could just swap the synccontent in/out instead of actually yielding and forcing to a thread pool thread. Have you considered this?

Reply:
We actually have our own sync context we use sometimes which posts to our IOThreadScheduler. The IOThreadScheduler hijacks an IO thread and uses that for dispatching work. The reason is we don't want to be affected by contention on the worker thread pool for things where bad stuff happens if you are late running code. For example, when you do an async IO operation you need a timer to cancel the IO as async IO doesn't honor timeouts. If your timer needs to be scheduled on the worker thread pool and it's contended (reached min threads and need to wait for another thread and there's a backlog of work), your timer will be late firing. This means a pending operation doesn't abort with an exception to user code on time. We've had customers open support issues because operations weren't timing out when they were supposed to and they were using timeouts in the double digit millisecond range. Another issue is when you make an async IO call such as a socket receive, the OS uses the originating thread and hijacks it to execute it's internal callback code when data has been received on a socket. The idea being the calling thread pays the CPU overhead and keeps scheduling fair. While there's a pending IO operation, you can't destroy the thread, so the threadpool has to keep the thread around even if there's no work for it to do. WCF by default uses a 10 minute receive timeout on our NetTcp transport, and it's configurable. I've seen customers set this to hours before. And then you have the OS thread scheduling behavior being different for IO threads than worker threads. Imagine you have a pending receive and some data for the socket arrives. The receive call was made on a worker thread so the OS kernel needs to hijack that thread. You have a lot of threads with high CPU contention so there's quite a delay until that thread gets scheduled so there's a delay before the IO completion callback happens telling you you've received some data. If you initiate the receive on an IO thread, the OS has different scheduling rules for IO threads so the hijacking will happen sooner and your incoming data is available sooner. It also allows the buffer to be unpinned sooner freeing up the gc. IO threads context switch a lot less. Anyway, the point being there are many reasons to schedule either IO or timers on an IO thread to improve performance and responsiveness and we achieve this by using a synchronization context. It would be difficult to reason about if any resetting of the sync context in any particular method would mess with our forcing IO onto an IO thread as changing a sync context also affects callers. We often don't have straight line code. We often call a method which spins off onto another thread and the existing thread then continues with other work.